### PR TITLE
Add GEMM device function

### PR DESCRIPTION
**Pull request by AGonzales-amd**
This adds a gemm device function which is callable in other kernels. The function is designed to be called by an entire wavefront and to compute a block of the output matrix. Therefore, problems can be decomposed into chunks that are operated on by individual wavefronts. Currently, it is used to implement an alternative rocsolver_gemm kernel.
 - Uses mfma instructions
 - Limited to `__gfx90a__`, `__gfx940__`, `__gfx941__`, and `__gfx942__` architectures
 - Supports both complex data types and transpose matrix operations

### CODE REVIEW

File: library/src/specialized/roclapack_gemm_device_functions.hpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+// Run with warpSize sized block
+template <typename T, typename I>
+__device__ void gemm_16x16xp(rocblas_operation transA,
+                                   rocblas_operation transB,
+                                   I m,
+                                   I n,
+                                   I p,
+                                   T alpha,
+                                   const T *A,
+                                   I inc_A,
```
Comment (EdDAzevedo) [inline library/src/specialized/roclapack_gemm_device_functions.hpp]: Is the inc_A, inc_B, ... etc the same as "stride" like strideA? If not, perhaps some description would be helpful.  Just a suggestion.

**Review by EdDAzevedo — APPROVED** [review]
It would be nice if there are tests scripts or updates to rocsolver-bench and/or rocsolver-test to check for correctness (and perhaps performance).  Note check for m,n,k to be nb, nb+1, nb-1  corner cases as well.

Will other BLAS operations, say TRMM, or SYR2K,   also take advantage of the new GEMM? One may consider a conceptually recursive formulation so that it can use GEMM.

File: library/src/specialized/roclapack_gemm_device_functions.hpp (hunk 1 above)
Comment (AGonzales-amd) [reply library/src/specialized/roclapack_gemm_device_functions.hpp]: Added function documentation, thanks for the suggestion

**Comment by jmachado-amd** [comment]
Hi @AGonzales-amd, I second Ed's suggestion: you should update `rocsolver-test` and `-bench` clients to support the internal gemm.  I can help you with the tests if you want (this would also be a good opportunity to provide a concrete answer to the question you asked in #879).

**Comment by AGonzales-amd** [comment]
> Hi @AGonzales-amd, I second Ed's suggestion: you should update `rocsolver-test` and `-bench` clients to support the internal gemm. I can help you with the tests if you want (this would also be a good opportunity to provide a concrete answer to the question you asked in #879).

Thanks @jmachado-amd, I could use your help. I did consider updating the client programs but I had trouble exporting the function or making it visible in the clients.

**Comment by AGonzales-amd** [comment]
Hi @jmachado-amd and @EdDAzevedo, the client programs have been updated to support internal gemm. One thing I'm not sure about is the tolerance for error checking.

**Comment by jmachado-amd** [comment]
Hi @AGonzales-amd, as long as the input matrices are "small" to "medium" sized and have positive, relatively small integer entries, the current test tolerance will work just fine! Let me know if you want to generalize the tests or just understand how those bounds work, and I can explain the important bits of the theory to you.

On another topic, I see that there are many `gemm` tests failing in Windows, you probably want to have a look at that sooner rather than later.

File: clients/gtest/unit/gemm_gtest.cpp [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+const vector<int64_t> large_n_size_range_64 = {1000};
+
+const vector<char> all_operations = {
+    'N',
+    'T',
+    'C',
+};
+
+const vector<double> alphas = {1.5};
+const vector<double> betas = {-2.0};
```
Comment (tfalders) [inline clients/gtest/unit/gemm_gtest.cpp]: I feel like alphas and betas can be combined into one vector.

File: clients/common/unit/testing_gemm.hpp [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+
+template <bool BATCHED, bool STRIDED, typename T, typename I>
+void testing_gemm(Arguments& argus)
+{
+    // get arguments
+    rocblas_local_handle handle;
+    rocblas_set_pointer_mode(handle, rocblas_pointer_mode_device);
+    I m = argus.get<I>("m");
+    I n = argus.get<I>("n");
+    I k = argus.get<I>("k");
```
Comment (tfalders) [inline clients/common/unit/testing_gemm.hpp]: I'd suggest setting n and k to default to m

File: clients/common/unit/testing_gemm.hpp [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+    I inca = argus.get<I>("inca", 1);
+    I incb = argus.get<I>("incb", 1);
+    I incc = argus.get<I>("incc", 1);
+
+    char tA = argus.get<char>("transA");
+    char tB = argus.get<char>("transB");
```
Comment (tfalders) [inline clients/common/unit/testing_gemm.hpp]: I'd suggest setting tA and tB to default to 'N'

File: library/src/specialized/roclapack_gemm_specialized_kernels.hpp [hunk 5]
```diff
@@ -201,41 +235,79 @@ rocblas_status rocsolver_gemm(rocblas_handle handle,
...
-                                strideB, beta, C, shiftC, incc, ldc, strideC);
+        const auto warpSize = deviceProperties.warpSize;
+
+        // launch specialized kernel
+        const I numWarpsX = 4;
+        const I numWarpsY = 4;
+        const I blocksx = (m + (numWarpsX * 16 - 1)) / (numWarpsX * 16);
+        const I blocksy = (n + (numWarpsY * 16 - 1)) / (numWarpsY * 16);
+        dim3 grid(blocksx, blocksy, batch_count);
+        dim3 threads(numWarpsX * warpSize, numWarpsY, 1);
```
Comment (tfalders) [inline library/src/specialized/roclapack_gemm_specialized_kernels.hpp:259]: This is a bit confusing for me. It looks like numWarpsY is actually the number of threads in y? What are numWarpsX * 16 and numWarpsY * 16 supposed to represent?

**Review by tfalders — APPROVED** [review]
The MFMA stuff is Greek to me, but everything else looks good. Just have some minor comments

File: library/src/specialized/roclapack_gemm_specialized_kernels.hpp (hunk 5 above)
Comment (EdDAzevedo) [reply library/src/specialized/roclapack_gemm_specialized_kernels.hpp:259]: Minor suggestion: it may be safer to query at execution time the actual value of "warpSize" and perhaps the max number of threads in a thread block.

Perhaps consider using a lambda  expression to calculate the ceiling function, say

auto ceil = [](auto n, auto b) { return( (n-1)/b + 1 ); };   

and use it as something like

const I blocksx = ceil( m,   (numWarpsX * 16) )

Comment (AGonzales-amd) [reply library/src/specialized/roclapack_gemm_specialized_kernels.hpp:259]: @tfalders, numWarps variables represent how many warps are working in parallel for each dimension. In this case, a thread group is laid out as 4x4 warps:
| | | | |
|---|---|---|----|
|w|w|w|w|
|w|w|w|w|
|w|w|w|w|
|w|w|w|w|

Each warp works on a 16x16 block of the output matrix so the m and n dimensions are ceil divided by numWarps * 16.

@EdDAzevedo I think a ceil function would be valuable and should be available everywhere. I think we should add this separately to change all ceil operations to a function call.

File: clients/common/unit/testing_gemm.hpp (hunk 4 above)
Comment (AGonzales-amd) [reply clients/common/unit/testing_gemm.hpp]: done

File: clients/common/unit/testing_gemm.hpp (hunk 3 above)
Comment (AGonzales-amd) [reply clients/common/unit/testing_gemm.hpp]: done

File: clients/gtest/unit/gemm_gtest.cpp (hunk 2 above)
Comment (AGonzales-amd) [reply clients/gtest/unit/gemm_gtest.cpp]: done

**Comment by tfalders** [comment]
Test failure in SYGVDX likely unrelated. Forcing the merge.
